WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListViewModel.kt

```diff
@@ -179,30 +181,100 @@ class PostListViewModel @Inject constructor(
-                } else {
-                    PostListEmptyViewState.EMPTY_LIST
-                }
-            } else {
-                PostListEmptyViewState.HIDDEN_LIST
-            }
+            createEmptyUiState()
         }
         result.addSource(pagedListWrapper.isEmpty) { result.value = update() }
         result.addSource(pagedListWrapper.isFetchingFirstPage) { result.value = update() }
         result.addSource(pagedListWrapper.listError) { result.value = update() }
         result
     }
 
+    private fun createEmptyUiState(): PostListEmptyUiState {
+        return if (pagedListWrapper.isEmpty.value != false) {
+            val error = pagedListWrapper.listError.value
+            val isLoadingFirstPage = pagedListWrapper.isFetchingFirstPage.value == true
+            when {
+                error != null -> createErrorListUiState(error)
+                isLoadingFirstPage -> PostListEmptyUiState.Loading
+                else -> createEmptyListUiState()
+            }
+        } else {
+            PostListEmptyUiState.HiddenList
+        }
+    }
+
+    private fun createErrorListUiState(error: ListError): PostListEmptyUiState {
+        return if (error.type == PERMISSION_ERROR) {
+            PostListEmptyUiState.PermissionsError
+        } else {
+            if (networkUtilsWrapper.isNetworkAvailable()) {
+                PostListEmptyUiState.RefreshError(R.string.error_refresh_posts)
+            } else {
+                PostListEmptyUiState.RefreshError(R.string.no_network_message)
+            }
+        }
+    }
+
+    private fun createEmptyListUiState(): PostListEmptyUiState.EmptyList {
+        return when (postListType) {
+            PUBLISHED -> PostListEmptyUiState.EmptyList(
+                    R.string.posts_published_empty,
+                    R.string.posts_empty_list_button,
+                    this@PostListViewModel::newPost
+            )
+            DRAFTS -> PostListEmptyUiState.EmptyList(
+                    R.string.posts_draft_empty,
+                    R.string.posts_empty_list_button,
+                    this@PostListViewModel::newPost
+            )
+            SCHEDULED -> PostListEmptyUiState.EmptyList(
+                    R.string.posts_scheduled_empty,
+                    R.string.posts_empty_list_button,
+                    this@PostListViewModel::newPost
+            )
+            TRASHED -> PostListEmptyUiState.EmptyList(R.string.posts_trashed_empty)
+        }
+    }
+
+    sealed class PostListEmptyUiState(
+        @StringRes val titleResId: Int? = null,
+        @DrawableRes val imgResId: Int? = null,
+        @StringRes val buttonTextResId: Int? = null,
+        val onButtonClick: (() -> Unit)? = null,
+        val listVisible: Boolean = true
+    ) {
+        class EmptyList(
+            titleResId: Int,
+            @StringRes buttonTextResId: Int? = null,
+            onButtonClick: (() -> Unit)? = null
+        ) : PostListEmptyUiState(
```

Comment:
Nitpick: Could we add the parameter names for the constructor calls in this `sealed class`?

Reply:
Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/9287/commits/9e1f8628c383e44886678f16378f9616f341825a